### PR TITLE
feat: added service worker with pure native implementation

### PR DESCRIPTION
**Pull request by 5war00p**
## What does this PR do?

Adds ServiceWorker to the web app by pure native implementaion (aka without any third party pacakges except browser's native and NodeJS).

## Related issues
/claim #3865
Close T-733

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (non-breaking small changes to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### <samp>🤖 Generated by Copilot at cfd69de</samp>

*  Add service worker support to web app to enable offline access and performance optimization
  * Create `ServiceWorker` component to register service worker file at root scope ([link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-a1361fd78f231f7812ccf9ce8b05bc4d806a115ba53c0f9d0ad1d80b5d065282R1-R15), [link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-a3569d521976a5ee9e97e2b2e005bf7423bc96bb49397b3adfb09cde5862bf93R6), [link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-a3569d521976a5ee9e97e2b2e005bf7423bc96bb49397b3adfb09cde5862bf93L13-R24))
  * Create custom script `build-sw.js` to bundle service worker code and inject static assets list as environment variable ([link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-4572ee9cc72ded9238b2de78cde9a4bf570e959358f22db209566e38c0d0eb8cR1-R26), [link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8L7-R9), [link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8R57), [link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8R99))
  * Create helper function `getFilesInDirectory` in `utils.js` to get list of static assets from `public` folder ([link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-8e33bbfdeb20f9b341a7a8f753da904e5fcb2875e5f2d81694840dd03a5a793dR1-R37))
  * Create `ServiceWorkerCache` class in `cache.ts` to encapsulate caching and serving logic using Cache API ([link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-3ff147c9c208c3f5bcac6bcbc995cf46f7805f6e123a38556c3645f11196957cR1-R112))
  * Create `index.ts` in `src/service-worker` folder as entry point for service worker code, which imports `ServiceWorkerCache` and registers event listeners for `fetch`, `install`, and `activate` events ([link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-70e3e8929a1a685f8319b9e64a1bbbbcf519eee33fbb0aa06072393963ca35f0R1-R45))
  * Add `Service-Worker-Allowed` header to `nextConfig` in `next.config.js` to allow service worker to control any URL under origin ([link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-a5de0fa0d80f14a44e9bdf2ed5d0cbb3a7bd5fdcacfbc1d03b5dcb57fc4943c1L67-R68))
  * Add `webworker` library to `tsconfig.json` to enable TypeScript recognition of service worker global scope and APIs ([link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-5ccca1924ddc8226881065b3da52870b4aa307c32dbacfc19eb22b8d5c8713f3L5-R11))
  * Add `service-worker.js` file to `.gitignore` and `ignorePatterns` of ESLint to exclude it from version control and linting ([link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-3a40410c8cd32e3860f5d9c6aafc7326d6a771ed4b01742fda42df2d4017871bL2-R3), [link](https://github.com/heyxyz/hey/pull/3876/files?diff=unified&w=0#diff-5fb16746c2a382d7958cc3954f65013c906c279eca2e2da671d22bb114ee301cL42-R42))

## Emoji

🚀🛠️🌐

### CODE REVIEW

File: apps/web/scripts/build-sw.js
Comment (bigint) [inline apps/web/scripts/build-sw.js:1]: rename scripts to script 🙇🏼

**Review by bigint — CHANGES_REQUESTED** [review]
Questions:

1. How this gonna build on vercel?
2. Will this cache CDN files too?

Suggestions

1. We can rename service-workers.js to sw.js
2. We can remove console.logs

File: apps/web/src/components/ServiceWorker/index.ts
Comment (5war00p) [reply apps/web/src/components/ServiceWorker/index.ts:4]: Yup, we can use useEffectOnce too, as we are doing only on first render

**Comment by 5war00p** [comment]
> Questions:
> 
> 1. How this gonna build on vercel?
> 2. Will this cache CDN files too?
> 
> Suggestions
> 
> 1. We can rename service-workers.js to sw.js
> 2. We can remove console.logs

Yes you can cache CDN files too, if you see in the `cache.ts` i'm checking for origin and calling cache function.
We can do the same for caching remote files by comparing url and maintaing list of CDN urls that needs to be cached.

**Comment by bigint** [comment]
Build fails in Vercel

**Comment by 5war00p** [comment]
Looks like the image you sent is broken, @bigint  can you resend?

**Comment by bigint** [comment]
image

**Comment by 5war00p** [comment]
mb, didn't changed script inside package.json, doing it now.

**Comment by 5war00p** [comment]
Looks like it failed again after a fix can you send me the screenshot again?

**Comment by bigint** [comment]
image

**Comment by 5war00p** [comment]
@bigint  looks like it's failed again and I see CI failing with some other code which I didn't touch (looks like main branch code).

**Comment by bigint** [comment]
Screenshot 2023-10-05 at 1 00 49 PM

@5war00p done force push, our git commit hook has lint and typecheck in it, you can manually check as well 🙇🏼

**Comment by 5war00p** [comment]
Yes, I'm running eslint and prettier but I'm contributing from a windows machine but with git bash, looks like these checks are not running properly in my machine.

**Comment by bigint** [comment]
@5war00p still the eslint fails 🙈 in local does it pass?

**Comment by 5war00p** [comment]
Yup it passes in local

**Comment by bigint** [comment]
You can check CI as well https://github.com/heyxyz/hey/actions/runs/6416001120/job/17418898953?pr=3876

**Comment by 5war00p** [comment]
if you see these changes are not mine:
image

**Comment by bigint** [comment]
This is weird 🙈

**Comment by bigint** [comment]
Screenshot 2023-10-05 at 2 37 26 PM

**Comment by 5war00p** [comment]
I figured out what might be the issue, pushing the fix now

**Comment by 5war00p** [comment]
image
It should be working now

**Comment by 5war00p** [comment]
Yo! @bigint vercel deployed now. Allow CI check so that will see if any other issues are there?

File: apps/web/script/build-sw.js
Comment (harish-sethuraman) [inline apps/web/script/build-sw.js:25]:
```suggestion
```

File: apps/web/scripts/build-sw.js
Comment (harish-sethuraman) [reply apps/web/scripts/build-sw.js:1]: LGTM

**Comment by bigint** [comment]
@5war00p I don't see SW is registered there

https://web-git-fork-5war00p-swaroop-eng-266-add-service-5f899f-heyxyz.vercel.app/

**Comment by 5war00p** [comment]
Looks like the public assets are not available on path
image
image
Browser recognizes there is a SW script file but its getting incorrect path

**Comment by 5war00p** [comment]
Hey @bigint , service worker installing now properly
image

image

**Comment by bigint** [comment]
Can we also configure CDN images?

**Comment by 5war00p** [comment]
Examples please?

**Comment by 5war00p** [comment]
@bigint Can you give CDN urls?

**Comment by bigint** [comment]
@5war00p well we need to cache all requests on the mentioned domains!

**Comment by 5war00p** [comment]
Okey, what's the domain list?

**Comment by 5war00p** [comment]
@bigint Configured cacheable domains too.
Here is the working screenshot:

image

**Comment by bigint** [comment]
Rest looks good 👯

**Comment by bigint** [comment]
Link T-733

**Comment by 5war00p** [comment]
@bigint resolved all the comments.

File: apps/web/package.json
Comment (bigint) [reply apps/web/package.json:9]: I still see this 🙈

File: apps/web/script/build-sw.mjs
Comment (bigint) [inline apps/web/script/build-sw.mjs:8]:
```suggestion
const STATIC_ASSETS = getFilesInDirectory('public/');
const outfile = 'public/sw.js'
```

**Review by bigint — CHANGES_REQUESTED** [review]

File: apps/web/package.json
Comment (bigint) [reply apps/web/package.json:9]: @5war00p can you reply instead of resolving multiple times?

File: apps/web/tsconfig.json
Comment (bigint) [inline apps/web/tsconfig.json:1]: this file is formated, fix this too!

**Comment by 5war00p** [comment]
@bigint fixed it, also one suggestion is can we have local prettier config so that we no need to worry about formatting every time. Irrespectivev of vscode extension settings, hey workspace always pose rules that mentioned in prettier config file

File: apps/web/package.json
Comment (bigint) [inline apps/web/package.json:9]: How about this?

Comment (bigint) [reply apps/web/package.json:9]: do we need build on dev?

**Comment by 5war00p** [comment]
@bigint  Let me know if you had any questions or need any other changes.

**Comment by bigint** [comment]
@5war00p https://github.com/heyxyz/hey/pull/3876#discussion_r1354104632

**Comment by 5war00p** [comment]
I gave reply to that
image

**Comment by 5war00p** [comment]
And why my replies are showing as pending?

**Comment by bigint** [comment]
I guess you are in review tab, you need to submit it!

**Comment by bigint** [comment]
@5war00p imo we can remove in dev command, it gonna create some problems while development!

**Comment by 5war00p** [comment]
Cool, okey. I will remove it.

**Comment by 5war00p** [comment]
@bigint Removed `build:sw` command from dev script

File: packages/config/eslint/base.js
Comment (bigint) [inline packages/config/eslint/base.js:42]: sw.js should be moved to web folder

**Review by bigint — CHANGES_REQUESTED** [review]

File: apps/web/src/sw/index.ts
Comment (bigint) [inline apps/web/src/sw/index.ts:38]: remove comments

File: apps/web/src/sw/index.ts
Comment (bigint) [inline apps/web/src/sw/index.ts:34]: don't need else statement, just directly return outside the if

File: apps/web/src/sw/cache.ts
Comment (bigint) [inline apps/web/src/sw/cache.ts:103]: still the else is here, we dont need it

**Comment by 5war00p** [comment]
I commented for those earlier, I'm having them as pending, dont know why. The else conditions I kept because the lint rules forced me to do so, one fo the vercel failing reasons is there is no else for if.

@bigint Anyway removed those, lets see the checks now. Also I gave one suggestion earlier and pasting again that, can we have local prettier config so that we no need to worry about formatting every time. Irrespectivev of vscode extension settings, hey workspace always pose rules that mentioned in prettier config file

**Comment by bigint** [comment]
Got blank page with console filled up with 404 after loading the site after long time 😕 

Also We need to wait till https://github.com/heyxyz/hey/pull/3892/ goes out 🙇🏼

**Comment by 5war00p** [comment]
Yes we can wait until dependent PR is merged.
Also can you share me the blank page issue screenshots or recordings?

**Comment by 5war00p** [comment]
@bigint  Any update on this effort?

**Comment by bigint** [comment]
@5war00p I see some issues with main URL caching, instead for now can we do one by one

- first cache and test subdomain assets
- next the main assets aka all js assets
- next we can implement the main doc URL caching (RN i see problem on this area!)

**Comment by bigint** [comment]
Closing this for now, you can raise the PR with updated one 🙇🏼

**Pull request closed without merging**
